### PR TITLE
Allow to skip the save post-step

### LINKED ISSUE #334: Support for Read Only Cache / Configurable Cache Saving
## Version

```
actions/cache@v2
```

## Description

We have a repository with multiple workflows that could benefit from shared caching across all of them. We would like to avoid trying to create and conditionalize one large workflow.

e.g. caching the download of all Go Modules:

```yaml
# Workflow 1
name: Code Checks

on:
  push:
    branches:
      - master
  pull_request:
    paths:
      - go.sum
      - src/**

jobs:
  go_mod_download:
    name: go mod download
    runs-on: ubuntu-latest
    steps:
    - uses: actions/checkout@v2
    - uses: actions/setup-go@v2
      with:
        go-version: "1.14"
    - uses: actions/cache@v2
      id: cache-go-pkg-mod
      with:
        path: ~/go/pkg/mod
        key: ${{ runner.os }}-go-pkg-mod-${{ hashFiles('go.sum') }}
    - if: steps.cache-go-pkg-mod.outputs.cache-hit != 'true' || steps.cache-go-pkg-mod.outcome == 'failure'
      run: go mod download
  # ... much more ...

# Workflow 2

name: CHANGELOG Checks
on:
  push:
    branches:
      - master
  pull_request:
    paths:
      - CHANGELOG.md

jobs:
  misspell:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v2
      - uses: actions/setup-go@v2
        with:
          go-version: "1.14"
      - uses: actions/cache@v2
        with:
          path: ~/go/pkg/mod
          key: ${{ runner.os }}-go-pkg-mod-${{ hashFiles('go.sum') }}
      - run: go install github.com/client9/misspell/cmd/misspell
      - run: misspell -error -source text CHANGELOG.md
```

Currently there can be race conditions between these workflows, where the first one that completes will save the cache, while other jobs will see the post-action hit on the cache key and prevent overwriting it. If the latter workflow completes first, the Go Modules download would contain only the download of the Go Modules relevant to the installed package, rather than all Go Modules for the project.

As a simple workaround, we could run the `go mod download` in every workflow:

```yaml
jobs:
  misspell:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v2
      - uses: actions/setup-go@v2
        with:
          go-version: "1.14"
      - uses: actions/cache@v2
        with:
          path: ~/go/pkg/mod
          key: ${{ runner.os }}-go-pkg-mod-${{ hashFiles('go.sum') }}
      - run: |
          go mod download
          go install github.com/client9/misspell/cmd/misspell
      - run: misspell -error -source text CHANGELOG.md
```

This could needlessly waste time/bandwidth, fail for unrelated networking reasons, and seems easy to miss though.

As a more complex workaround, we could add the `go_mod_download` job to every workflow:

```yaml
jobs:
  go_mod_download:
    name: go mod download
    runs-on: ubuntu-latest
    steps:
    - uses: actions/checkout@v2
    - uses: actions/setup-go@v2
      with:
        go-version: "1.14"
    - uses: actions/cache@v2
      id: cache-go-pkg-mod
      with:
        path: ~/go/pkg/mod
        key: ${{ runner.os }}-go-pkg-mod-${{ hashFiles('go.sum') }}
    - if: steps.cache-go-pkg-mod.outputs.cache-hit != 'true' || steps.cache-go-pkg-mod.outcome == 'failure'
      run: go mod download
  misspell:
    needs: [go_mod_download]
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v2
      - uses: actions/setup-go@v2
        with:
          go-version: "1.14"
      - uses: actions/cache@v2
        with:
          path: ~/go/pkg/mod
          key: ${{ runner.os }}-go-pkg-mod-${{ hashFiles('go.sum') }}
      - run: go install github.com/client9/misspell/cmd/misspell
      - run: misspell -error -source text CHANGELOG.md
```

But this feels like a lot of overhead and maintenance for each workflow. It would be great if we could tell this action to skip the post save action, effectively making the cache best effort and read-only, while preventing it from missing contents.

## Proposals

Adding a `read-only` input:

```yaml
jobs:
  misspell:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v2
      - uses: actions/setup-go@v2
        with:
          go-version: "1.14"
      - uses: actions/cache@v2
        with:
          path: ~/go/pkg/mod
          key: ${{ runner.os }}-go-pkg-mod-${{ hashFiles('go.sum') }}
          read-only: true
      - run: go install github.com/client9/misspell/cmd/misspell
      - run: misspell -error -source text CHANGELOG.md
```

Or maybe adding a `save` input (see references for why this may be beneficial):

```yaml
jobs:
  misspell:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v2
      - uses: actions/setup-go@v2
        with:
          go-version: "1.14"
      - uses: actions/cache@v2
        with:
          path: ~/go/pkg/mod
          key: ${{ runner.os }}-go-pkg-mod-${{ hashFiles('go.sum') }}
          save: never # or "always", defaults to current "when missing"
      - run: go install github.com/client9/misspell/cmd/misspell
      - run: misspell -error -source text CHANGELOG.md
```

Thank you for the consideration! Please reach out if I'm missing something and this is already possible or if this would be acceptable as an enhancement.

## References

- Configurable save cache on failure: https://github.com/actions/cache/issues/92
- Enable caching of dependencies even if builds fail with post-if modifier: https://github.com/actions/cache/issues/165

### LINKED ISSUE #350: Read only cache?
I am interested in creating cache in action X of the workflow, and then reusing it in action Y, but I don't want to waste time to _recreate_ cache at the cleanup stage of the `cache` action in action Y  - I know I only used it in read-only fashion in action Y (whereas action X might have written to it), so waiting for the new cache creation is wasteful.

Is there a way in current implementation to get this behavior and if not would you consider this as a new feature?

### PR DESCRIPTION
**Pull request by roni-frantchi**
This is similar to #474 - but improves on that by skipping the `post` (save) step entirely, saving tens of seconds on each run.  

Fixes #350 
Fixes #334

### CODE REVIEW

**Comment by roni-frantchi** [comment]
Thanks @domjtalbot.  
Sure thing.  
If we get a positive response from maintainers on pushing it in, I'd be happy to make that adjustment along with any others suggested.

File: README.md
Comment (joshmgross) [inline README.md:84]: This would need to be an `env` variable on the `cache-primes` step

File: action.yml
Comment (joshmgross) [inline action.yml:24]: Unfortunately this doesn't work well, as `env` values are always strings.

So for example both `CACHE_SKIP_SAVE: true` and `CACHE_SKIP_SAVE: false` would evaluate to a non-empty string ('true' and 'false`) which equal `true`.

See the debug logging in https://github.com/joshmgross/actions-testing/runs/1560134417?check_suite_focus=true

**Review by joshmgross — COMMENTED** [review]
Thanks for your suggestion!

Due to the type issues with `env` always being a string, I think we'd need this to be an input rather than an environment variable so that we can properly evaluate it.

Also this does not make the cache key option, so could you update your PR description to remove "Fixes #210"?

File: action.yml
Comment (roni-frantchi) [reply action.yml]: Right!  
Thanks.  

How about that then -https://github.com/actions/cache/pull/489/commits/f4a5fe8db9227c95e5652eec8e065832f1b4b4d9

File: README.md
Comment (roni-frantchi) [reply README.md:86]: https://github.com/actions/cache/pull/489/commits/6ceddd55d128082f23eaf7026dfd38e5f14baa41

**Review by roni-frantchi — COMMENTED** [review]
Thanks for reviewing @joshmgross !  

> I think we'd need this to be an input rather than an environment variable so that we can properly evaluate it

`post-if` is undocumented and despite my best efforts to make it evaluate inputs I couldn't - only `env` was available to me.  

> Due to the type issues with env always being a string...

I did address this one though - so perhaps using `env` would be acceptable now?

**Comment by anders-kiaer** [comment]
I see this PR has gone through a first review by @joshmgross already. Will it be merged in the near future, or should use cases needing to skip the save step use e.g. a fork of this action?

Regarding using environment variable: I think this is a good choice, if my understanding is correct that previous steps in the job then can dynamically/during runtime instruct the `cache` action save/post step through https://docs.github.com/en/actions/reference/workflow-commands-for-github-actions#setting-an-environment-variable.

Thanks :slightly_smiling_face:

**Comment by mikeage** [comment]
This looks wonderful; one question, if you don't mind. I have a very large cache (~750 MB), and when I have multiple PRs in progress, it appears that the 5GB cache gets filled up with those, and the cache from `main` gets pushed out. If I were to use this fork and set the option to skip the saving on the PRs but not on the main branch, this would keep the main cache as a base for the all of the PRs, right?

**Comment by zwass** [comment]
@joshmgross Can you please re-review/merge this? I'd like to use it and would prefer not to rely on a fork.

**Comment by joshmgross** [comment]
👋  Hey @zwass, the @actions/artifacts-actions team should be able to help out here.

File: README.md
Comment (dhadka) [inline README.md:45]:
```suggestion
* `CACHE_SKIP_SAVE` - [optional] When set to `true`, any modifications made to the restored cache will not be persisted back at the end of the step.  This environment variable can be set at any time using `echo "CACHE_SKIP_SAVE=true" >> $GITHUB_ENV`
```

File: README.md
Comment (dhadka) [inline README.md:74]: Given that this is the default example and skipping cache save is a special case, I wouldn't include this here.  Perhaps move this into the Examples.md file and we can add some specific examples of where this is useful.

File: README.md
Comment (dhadka) [inline README.md:87]: This should be removed now that it's set on the `cache-primes` step.

File: README.md
Comment (dhadka) [reply README.md:74]: Something we can also highlight in the example is how users can either set this on the cache step as shown here or using `echo "CACHE_SKIP_SAVE=true" >> $GITHUB_ENV`.

**Review by dhadka — CHANGES_REQUESTED** [review]
TY for proposing this change.  Just some minor comments on the README changes, but otherwise looks good.  Please ping me when ready and I can 🚢 this.

File: README.md
Comment (roni-frantchi) [inline README.md:87]:
```suggestion
```

File: README.md
Comment (roni-frantchi) [reply README.md]: https://github.com/actions/cache/pull/489/commits/91185bb10f818be7d9890148bf120e8c89440cc0

File: README.md
Comment (roni-frantchi) [reply README.md:74]: I think the changes you've suggested on the readme file capture it well enough

File: README.md
Comment (roni-frantchi) [reply README.md]: https://github.com/actions/cache/pull/489/commits/64daede5552c68991cba51f3bc0ac2bc26945a11

**Comment by ollydev** [comment]
@dhadka This is what I wanted #498 for, didn't see this PR though.

Right now cache will never be saved if the job fails.
Example: Spend 10mins building dependencies (succesfully) but a future job (such as failing a test) fails. Cache will not save!

Can we add something like:
```yml
env:  
  CACHE_FORCE_SAVE: true
```
```yml
`post-if: success() && env.CACHE_SKIP_SAVE != 'true' || failed() && env.CACHE_FORCE_SAVE!= 'true'`
```

**Comment by zwass** [comment]
Perhaps @ollydev's request could be merged into this one with a single flag?

`CACHE_SAVE=always` - Always save regardless of errors in preceding steps.
`CACHE_SAVE=never` - Never save regardless of errors.
Not set - Default behavior (as before this PR).

**Comment by jmoguillansky-gpsw** [comment]
failed() && env.CACHE_FORCE_SAVE == 'true': how cached content was generated successfully?  I think we need some mechanism to indicate that generating the cache content was successful, and then cache it immediately, why wait until the end of the workflow?

**Comment by bgianfo** [comment]
>  Please ping me when ready and I can 🚢 this.

It looks like all the comments have been addressed?

**Comment by bgianfo** [comment]
> Please ping me when ready and I can 🚢 this.

@dhadka Looks like all the comments have been addressed, can we get this merged?

**Comment by dhadka** [comment]
@bgianfo There's a discussion happening over in https://github.com/actions/cache/pull/571 on a similar issue.  Essentially, there are a few open issues that are somewhat related (better control over if and when the cache is saved).  Instead of adding in a bunch of piecemeal fixes, is there a more generalizable solution?

> failed() && env.CACHE_FORCE_SAVE == 'true': how do we know the content to cache was generated successfully? I think we need some mechanism to indicate that generating the cache content was successful, and then cache it immediately, why wait until the end of the workflow?

@jmoguillansky-gpsw Yes, this is essentially a limitation of running save as a "post step", which only runs at the end of the workflow.  You're essentially looking for a standalone save step that a user can run at any point in their workflow.  See https://github.com/actions/cache/pull/571#issuecomment-839123699.

**Comment by Zoramite** [comment]
How does this work when you are using multiple cache actions in one workflow job?

I have a cache action for the caching node modules and I have another cache action for a benchmark build results.

If I understand this correctly, when I `echo "CACHE_SKIP_SAVE=true" >> $GITHUB_ENV` then it would prevent both the cache for the node modules **and** the cache for the benchmark to not save after the workflow is completed. Is that correct?  I really want to target one of the cache actions to not save (ex: don't save the benchmark cache unless the branch is `main`), not just blindly affect all of the cache saves.

It seems like there should be a better way for specifically preventing a single cache action from saving without **all** cache actions being prevented from caching?

**Comment by neerfri** [comment]
@dakale @roni-frantchi Seems like this change is of interest to many users.
As we are not bringing this PR to conclusion downstream users are turning to use forks.
Would probably be a good idea to bring this to conclusion

Thanks!

**Comment by roni-frantchi** [comment]
@dakale I've addressed all of you comments - is there anything other request in this thread that you think must be taken care of in this PR before merging?  

I would rather add any improvements later on and get this one going.

File: README.md
Comment (rene-leanix) [reply README.md:74]: @roni-frantchi & @dhadka, would be great if this PR could be merged to reduce the workflow runtime when a cache-update isn't required. Seems like only a small adjustment would be needed?
Not sure if [this issue](https://github.com/actions/cache/issues/597) could be addressed in this PR too: skipping the post-step altogether when there was a cache-hit.

**Comment by vinaysshenoy** [comment]
Is this going to get merged soon? The ability to skip the save post-step is something that will be very useful for our use case as well.

File: README.md
Comment (roni-frantchi) [reply README.md:74]: > Not sure if this issue could be addressed in this PR too

I'm not sure that it can either...  
Seeing so many are expecting it, maybe we should explore doing it on another PR?..  

> Seems like only a small adjustment would be needed?

Sorry @rene-leanix it's been a while - what adjustment would that be?..

File: action.yml
Comment (JCMais) [inline action.yml:24]: My only suggestion would be to have a more specific environment variable here.

Something like `GH_ACTIONS_CACHE_SKIP_SAVE`

File: README.md
Comment (rene-leanix) [reply README.md:74]: @roni-frantchi, I was just referring to @dhadka's comments above in this thread: [here](https://github.com/actions/cache/pull/489/files#r616281599) and [here](https://github.com/actions/cache/pull/489/files#r616282355).

**Comment by Sytten** [comment]
Any hope to get this merged

**Comment by iuioiua** [comment]
@roni-frantchi, is it possible for you to update the branch? It'd remove the deprecation warnings around save-state, set-output and Node.js 12 for those using the fork.

**Comment by Jimimaku** [comment]
> Dies ist ähnlich wie [# 474](https://github.com/actions/cache/pull/474) - verbessert dies jedoch, indem der (gespeicherte) Schritt vollständig übersprungen wird, wodurch bei jedem Lauf Dutzende von Sekunden gespart werden.`post`
> 
> Behebt #350 Behebt #334

**Comment by kotewar** [comment]
Hey all, 👋🏽 

We have created a [discussion](https://github.com/actions/cache/discussions/1020) with a proposal that will solve this problem, do let us know your feedback, thanks 😊

**Comment by kotewar** [comment]
Hey everyone 👋🏽

Two new actions [actions/cache/restore](https://github.com/actions/cache/tree/main/restore) and [actions/cache/save](https://github.com/actions/cache/tree/main/save) are now available with tag `v3` to everyone for use. These can now be used to achieve granular control on the restore and save steps in the cache action. 

Do try them and let us know your feedback. We hope these new actions will take care of your use cases. 🙇🏽 

Closing this PR now as we believe the new actions will take care of the same, feel free to reopen it or create an issue if need be.

**Pull request closed without merging**
